### ISSUE TITLE
[Incorrect gradients] `_is_perturbed` produces the wrong value when used inside `lax.scan` etc.

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#9567]
This:
```python
import jax
import jax.lax as lax
from jax._src.custom_derivatives import _is_perturbed

def f(x):
    def g(y, _):
        z = y * x
        print(_is_perturbed(z))
        return y, None
    g(1, None)
    return lax.scan(g, 1, xs=None, length=1)[0]

jax.jvp(f, (1.0,), (1.0,))
```
produces:
```
True
False
```

The correct behaviour is for both to be `True`.

I imagine this is silently setting a few gradients to zero, somewhere out there.

**Comment by mattjj** [comment]
We're going to revert #9923 so I'll reopen this for now...

**Comment by mattjj** [comment]
Actually, the current plan is to leave it merged, though there's more to follow up on this front...
